### ISSUE TITLE
FastLED: Seperate out Global Brightness from Component Scale

### ISSUE THREAD

**Issue by zackees** [FastLED/FastLED#1709]
This is a brain dump on what it takes to fix the APA102 global brightness issue, and all other led chipsets with higher resolution than RGB8.

TL;DR: Brightness and component scaling factors are pre-mixed. This is sufficient for RGB8 chipsets. But for higher precision chipsets like the APA102, this results in a reduction in color resolution for lower global brightness levels. This feature request will specify a way to fix this by storing additional brightness and unmixed component scales so higher precision chipsets can more intelligently mix them together at the driver level.

## Background

PixelController bakes in the component brightness with global brightness and sends this to the Clocked and Clockless device drivers.

The pixel controller takes as input the led color scale/temperature pre-mixed with the global brightness.

This made sense for RGB8 led chipsets, with limited resolution. However, for the newer chipsets this results in a loss of color accuracy.

For example, the APA102 chipset has RGB8 + 5-bit brightness. When I made the APA102HD mode with 5bit brightness gamma correction I first encountered this problem of how to apply global brightness.

For example, this is the APA102's code for handling global brightness:

```C++
	static inline void getGlobalBrightnessAndScalingFactors(
		    PixelController<RGB_ORDER>& pixels,
		    uint8_t* out_s0, uint8_t* out_s1, uint8_t* out_s2, uint8_t* out_brightness) {
		uint8_t s0 = pixels.getScale0();
		uint8_t s1 = pixels.getScale1();
		uint8_t s2 = pixels.getScale2();
#if FASTLED_USE_GLOBAL_BRIGHTNESS == 1
		const uint16_t maxBrightness = 0x1F;
		uint16_t brightness = ((((uint16_t)max(max(s0, s1), s2) + 1) * maxBrightness - 1) >> 8) + 1;
		s0 = (maxBrightness * s0 + (brightness >> 1)) / brightness;
		s1 = (maxBrightness * s1 + (brightness >> 1)) / brightness;
		s2 = (maxBrightness * s2 + (brightness >> 1)) / brightness;
#else
		const uint8_t brightness = 0x1F;
#endif
		*out_s0 = s0;
		*out_s1 = s1;
		*out_s2 = s2;
		*out_brightness = static_cast<uint8_t>(brightness);
	}
```

This code appears to be doing a complicated re-normalization of the color components.

Ideally this code should be as simple as this:

```C++
uint8_t s0 = pixels.getScale0();
uint8_t s1 = pixels.getScale1();
uint8_t s2 = pixels.getScale2();
uint8_t brightness_5bit = map(brightness_8bit, 0, 0xff, 0, 0x1f);
```

### plan of action

The changes necessary for this refactor include the following:

  * PixelController will add two additional data members:
    * CRGB mComponentScale
    * uint8_t mGlobalBrightness

The pre-mixed mScale variable, which includes brightness mixed in with component scale, will remain in the PIxelController and continue to work like normal. Chipset drivers like the APA102 will switch from using the pre-mixed component scale to the mComponentScale to color adjust the leds and mGlobalBrightness to set the 5-bit global brightness value.

# Future Enhancements

### UCS7604

This is an LED chipset that has an optional RGB16 mode. Global brightness can be optimally mixed only in this mode.

In RGB8 mode, this chipset will use the pre-mixed scaling factors in the same way as WS2812 current does today.

In RGB16 mode, this chipset will employ the following code:

```C++
uint16_t map16(uint8_t x) { return map(x, 0, 0xff, 0, 0xffff);

uint16_t r16 = map16(r);
uint16_t g16 = map16(g);
uint16_t b16 = map16(b);
uint16_t global_brightness16 = map16(global_brightness);

r16 = scale(r16, global_brightness16);
g16 = scale(g16, global_brightness16);
b16 = scale(b16, global_brightness16);

// Note that UCSS7604 always does gamma correction in RGB8 mode at gamma = 2.2, according to the datasheet.
// So this has to be done here in software.
r16 = gamma_correct(r16);
g16 = gamma_correct(g16);
b16 = gamma_correct(b16);

// Component scale is a linear power adjust feature so must come AFTER gamma correction.
// Also, it's possible because of the internal LUT of this chipset that component scale factors
// may not even be necessary to color balance the strip and thus may be ignore entirely.
if (UCS7604_SHOULD_APPLY_COLOR_BALANCE) {
  r16 = scale(r16, mComponentScale.r);
  g16 = scale(g16, mComponentScale.g);
  b16 = scale(b16, mComponentScale.b);
}

driver.sendPixel(r16, g16, b16);
```

Unfortunately the cost of using this higher precision global brightness for the UCS7604 chipset will be require that double the byte payload will have to be sent out to the chipset, halving the framerate. It will probably also not be enabled for the older AVR chipsets which have very tight timing requirements.

**Comment by zackees** [comment]
This is now done and in FastLED 3.8 beta.

LOOKS FANTASTIC! Global Brightness is now default on starting in 3.8.
